File: backend/services/speech_normalize.py
"""Normalize spoken punctuation into real symbols for voice transcripts."""

import re

_AT_PATTERNS = [
    re.compile(r"\bat\s*the\s*rate\b", re.I),
    re.compile(r"\battherate\b", re.I),
    re.compile(r"\bat\s*sign\b", re.I),
    re.compile(r"\bat\s*symbol\b", re.I),
]

_TLD = re.compile(r"\b(com|org|net|io|co|in|edu|gov|uk|au|ca|me|dev|ai)\b", re.I)
_PROVIDER = re.compile(r"\b(gmail|yahoo|outlook|hotmail|icloud|protonmail|company)\b", re.I)
_LONE_AT = re.compile(r"(\w[\w.\-]*)\s+at\s+(\w[\w.\-]*)", re.I)


def _should_normalize_dots(text: str) -> bool:
    if "@" in text:
        return True
    if len(re.findall(r"\bdot\b", text, re.I)) >= 2:
        return True
    if re.search(r"\b\w+\s+dot\s+(com|org|net|io|co|in|edu|gov)\b", text, re.I):
        return True
    if re.search(r"\bat\s+(gmail|yahoo|outlook|hotmail|company)\b", text, re.I):
        return True
    return False


def normalize_spoken_symbols(text: str) -> str:
    """e.g. 'ronit at gmail dot com' → 'ronit@gmail.com'"""
    if not text:
        return text

    t = text.strip()

    for pat in _AT_PATTERNS:
        t = pat.sub("@", t)

    def _lone_at(m: re.Match) -> str:
        before, after = m.group(1), m.group(2)
        if _PROVIDER.search(after) or _TLD.search(after) or "." in before:
            return f"{before}@{after}"
        return m.group(0)

    t = _LONE_AT.sub(_lone_at, t)

    if _should_normalize_dots(t):
        t = re.sub(r"\s+dot\s+", ".", t, flags=re.I)
        t = re.sub(r"\bdot\b(?=\s*[\w])", ".", t, flags=re.I)

    if _should_normalize_dots(t) and re.search(r"\bperiod\b", t, re.I):
        t = re.sub(r"\s+period\s+", ".", t, flags=re.I)

    t = re.sub(r"\s*@\s*", "@", t)
    t = re.sub(r"\s*\.\s*", ".", t)

    t = re.sub(r"\s+underscore\s+", "_", t, flags=re.I)
    t = re.sub(r"\s+hyphen\s+", "-", t, flags=re.I)
    t = re.sub(r"\s+dash\s+", "-", t, flags=re.I)
    t = re.sub(r"\s+slash\s+", "/", t, flags=re.I)
    t = re.sub(r"(\d)\s+dot\s+(\d)", r"\1.\2", t, flags=re.I)

    t = re.sub(r"\s{2,}", " ", t).strip()
    return t

File: frontend/src/utils/speechNormalize.js
/**
 * Convert spoken punctuation/symbols into real characters.
 * e.g. "ronit at gmail dot com" → "ronit@gmail.com"
 */

const AT_PATTERNS = [
  /\bat\s*the\s*rate\b/gi,
  /\battherate\b/gi,
  /\bat\s*sign\b/gi,
  /\bat\s*symbol\b/gi,
];

const TLD_PATTERN = /\b(com|org|net|io|co|in|edu|gov|uk|au|ca|me|dev|ai)\b/i;
const EMAIL_PROVIDER = /\b(gmail|yahoo|outlook|hotmail|icloud|protonmail|company)\b/i;

function shouldNormalizeDots(text) {
  if (text.includes('@')) return true;
  const dotCount = (text.match(/\bdot\b/gi) || []).length;
  if (dotCount >= 2) return true;
  if (/\b\w+\s+dot\s+(com|org|net|io|co|in|edu|gov)\b/i.test(text)) return true;
  if (/\bat\s+(gmail|yahoo|outlook|hotmail|company)\b/i.test(text)) return true;
  if (/\bperiod\b/i.test(text) && dotCount >= 1) return true;
  return false;
}

function shouldNormalizePeriod(text) {
  return shouldNormalizeDots(text) || (text.match(/\bperiod\b/gi) || []).length >= 2;
}

/**
 * @param {string} text
 * @returns {string}
 */
export function normalizeSpokenSymbols(text) {
  if (!text || typeof text !== 'string') return text;

  let t = text.trim();

  for (const pat of AT_PATTERNS) {
    t = t.replace(pat, '@');
  }

  // Lone "at" between email-like tokens: "john at gmail"
  t = t.replace(
    /(\w[\w.-]*)\s+at\s+(\w[\w.-]*)/gi,
    (match, before, after) => {
      if (EMAIL_PROVIDER.test(after) || TLD_PATTERN.test(after) || before.includes('.')) {
        return `${before}@${after}`;
      }
      return match;
    }
  );

  if (shouldNormalizeDots(t)) {
    t = t.replace(/\s+dot\s+/gi, '.');
    t = t.replace(/\bdot\b(?=\s*[\w])/gi, '.');
  }

  if (shouldNormalizePeriod(t)) {
    t = t.replace(/\s+period\s+/gi, '.');
  }

  t = t.replace(/\s*@\s*/g, '@');
  t = t.replace(/\s*\.\s*/g, '.');

  t = t.replace(/\s+underscore\s+/gi, '_');
  t = t.replace(/\s+hyphen\s+/gi, '-');
  t = t.replace(/\s+dash\s+/gi, '-');
  t = t.replace(/\s+slash\s+/gi, '/');

  // "3 dot 30" → "3.30" for times
  t = t.replace(/(\d)\s+dot\s+(\d)/gi, '$1.$2');

  t = t.replace(/\s{2,}/g, ' ').trim();
  return t;
}
